src/utils/first_follow.cpp

```diff
@@ -0,0 +1,267 @@
+      // if production head / parent is the start_symbol, then FOLLOW(head) = {$}
+      // $ -> input endmarker
+      follows.insert(make_pair(production.GetParent(), std::vector<std::string>(1, STRING_ENDMARKER)));
+    } else {
+      follows.insert(make_pair(production.GetParent(), std::vector<std::string>(0)));
+    }
+  }
+
+  std::function<void(const std::string &)> calc_recursive;
+  // lambda function that returns nothing
+  calc_recursive = [&](const std::string &key) {
+    for (const auto &rule : GetRulesForParent(augmented_grammar, key)) {
+      std::vector<std::string>::const_iterator symbol_itr;
+      std::vector<std::string>::const_iterator next_itr;
+      for (symbol_itr = rule.GetEntities().begin(); symbol_itr != rule.GetEntities().end(); symbol_itr++) {
+        std::string mid = *symbol_itr;
+        // if A -> alpha B beta is a production, where B -> non-terminal & alpha, beta -> set of symbols
+        // then FOLLOW(B) contains {FIRST(beta) - EPSILON}
+        // if mid aka current symbol is terminal, then ignore
+        if (SearchParent(augmented_grammar, mid)) {
+          for (next_itr = symbol_itr + 1; next_itr != rule.GetEntities().end(); next_itr++) {
+            // next_itr -> iterates through every symbol in beta
+            for (const auto &der : firsts[*next_itr]) {
+              // discarding EPSILON
+              if (der != std::string(grammar::EPSILON) &&
+                  find(follows[mid].begin(), follows[mid].end(), der) == follows[mid].end()) {
+                follows[mid].push_back(der);
+                finished = false;
+              }
+            }
+
+            // non-nullable symbol encountered, so no need of deriving succeeding symbols
+            if (!nullables.at(*next_itr)) {
+              break;
+            }
+          }
+
+          /* if execution has reached this point, then beta => EPSILON */
+
+          // if A -> alpha B / A -> alpha B beta & beta => EPSILON, then
+          // FOLLOW(B) includes FOLLOW(A)
+          if (next_itr == rule.GetEntities().end()) {
+            for (const auto &der : follows[key]) {
+              if (find(follows[mid].begin(), follows[mid].end(), der) == follows[mid].end()) {
+                follows[mid].push_back(der);
+                finished = false;
+              }
+            }
+          }
+        }
+      }
+    }
+  };
+
+  // boolean variable finished used for same purpose as in CalcFirsts
+  while (!finished) {
+    finished = true;
+    calc_recursive(start_symbol);
+    for (const auto &production : augmented_grammar) {
+      // FOLLOW(start_symbol) remains unchanged, and is always {$}
+      if (production.GetParent() != start_symbol) {
+        calc_recursive(production.GetParent());
+      }
+    }
+  }
+
+  // sorting FOLLOW mappings for convenience in presentation
+  for (const auto &production : augmented_grammar) {
+    sort(follows[production.GetParent()].begin(), follows[production.GetParent()].end());
+  }
+
```

Comment:
This step is redundant and will add extra to the compilation time.

Reply:
This makes easier to write tests.

Reply:
That is not a valid reason.